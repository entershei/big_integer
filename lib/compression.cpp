#include <vector>
#include <algorithm>
#include <iostream>
#include <set>
#include <cstring>
#include "compression.h"

using char_t = unsigned char;

const size_t ALPHABET = 256;

struct buffer_input_stream {
    static constexpr size_t BUFFER_SIZE = 1U << 15;
    char buffer[BUFFER_SIZE];
    size_t pos;
    std::istream &in;
    size_t size;
    size_t pos_in_char;
    uint64_t cnt_in_bit;

    explicit buffer_input_stream(std::istream &in) : in(in), pos_in_char(0), cnt_in_bit(0) {
        buf_update();
    }

    void buf_update() {
        pos = 0;
        in.read(buffer, BUFFER_SIZE);
        size = static_cast<size_t >(in.gcount());
        pos_in_char = 0;
    }

    bool can_read_char() {
        if (pos < size) { return true; }

        buf_update();
        return pos < size;
    }

    bool read_char(char_t &symbol) {
        if (!can_read_char()) { return false; }
        symbol = static_cast<char_t>(buffer[pos]);
        ++pos;
        pos_in_char = 0;

        return true;
    }

    bool read_bit(bool &bit) {
        if (pos == size) {
            buf_update();
        }
        if (pos == size) {
            return false;
        }

        bit = static_cast<bool>(static_cast<char_t>(buffer[pos]) & (1U << (7 - pos_in_char)));
        ++pos_in_char;

        if (pos_in_char == 8) {
            pos_in_char = 0;
            ++pos;
        }

        return true;
    }

    void refresh() {
        in.clear();
        in.seekg(0);
    }
};

struct buffer_output_stream {
    static constexpr size_t BUFFER_SIZE = 1U << 15;
    char buffer[BUFFER_SIZE];
    size_t pos;
    std::ostream &out;
    size_t pos_in_char;

    buffer_output_stream(std::ostream &out) : pos(0), out(out), pos_in_char(0) {
        memset(buffer, 0, sizeof buffer);
    }

    void write_from_buffer() {
        out.write(buffer, pos);
        memset(buffer, 0, sizeof buffer);
        pos = 0;
        pos_in_char = 0;
    }

    void write_char(char symbol) {
        if (pos == BUFFER_SIZE) {
            write_from_buffer();
        }

        buffer[pos] = symbol;
        ++pos;
        pos_in_char = 0;
    }

    void write_bit(bool bit) {
        if (pos == BUFFER_SIZE) {
            write_from_buffer();
        }

        buffer[pos] |= bit << (7 - pos_in_char);
        ++pos_in_char;

        if (pos_in_char == 8) {
            pos_in_char = 0;
            ++pos;
        }
    }

    ~buffer_output_stream() {
        if (pos_in_char) {
            ++pos;
        }
        write_from_buffer();
    }
};

std::vector<size_t> count_frequency(buffer_input_stream &bin) {
    std::vector<size_t> frequency(ALPHABET);
    char_t symbol;

    while (bin.read_char(symbol)) {
        ++frequency[symbol];
    }

    bin.refresh();

    return frequency;
}

struct node {
    int left;
    int right;

    node() : left(-1), right(-1) {}

    node(int left, int right) : left(left), right(right) {}
};

bool cmp(std::pair<size_t, size_t> const &a, std::pair<size_t, size_t> const &b) {
    return a.second < b.second;
}

std::vector<node> build_huffman_tree(std::vector<size_t> const &frequency) {
    std::vector<node> tree(ALPHABET * 2 - 1);
    std::set<std::pair<size_t, size_t> > freq; //второе - индекс в tree

    for (size_t i = 0; i < ALPHABET; ++i) {
        freq.insert({frequency[i], i});
    }

    size_t cnt_nodes = ALPHABET;
    while (freq.size() > 1) {
        auto node1 = *freq.begin();
        freq.erase(freq.begin());
        auto node2 = *freq.begin();
        freq.erase(freq.begin());
        tree[cnt_nodes] = {static_cast<int>(node1.second), static_cast<int>(node2.second)};
        freq.insert({node1.first + node2.first, cnt_nodes});
        ++cnt_nodes;
    }

    return tree;
}

void count_length(size_t pos, char_t depth, std::vector<node> const &tree, std::vector<char_t> &length) {
    if (tree[pos].left == -1 && tree[pos].right == -1) {
        length[pos] = depth;
        return;
    }

    if (tree[pos].left != -1) {
        count_length(static_cast<size_t>(tree[pos].left), depth + 1, tree, length);
    }
    if (tree[pos].right != -1) {
        count_length(static_cast<size_t>(tree[pos].right), depth + 1, tree, length);
    }
}

std::vector<size_t> count_key(std::vector<char_t> const &len) {
    std::vector<std::pair<size_t, size_t> > length(ALPHABET);

    for (size_t i = 0; i < ALPHABET; ++i) {
        length[i] = {i, len[i]};
    }

    sort(length.begin(), length.end(), cmp);

    std::vector<size_t> code(ALPHABET);
    size_t cur_code = 0;
    code[length[0].first] = 0;

    for (size_t i = 1; i < ALPHABET; ++i) {
        ++cur_code;

        if (length[i - 1].second < length[i].second) {
            cur_code <<= (length[i].second - length[i - 1].second);
        }

        code[length[i].first] = cur_code;
    }

    return code;
}

void write_cnt_bit(buffer_output_stream &out, uint64_t cnt) {
    const size_t BASE = (1U << 8) - 1;
    for (size_t i = 8; i-- > 0;) {
        out.write_char(static_cast<char_t>((cnt >> (i * 8)) & BASE));
    }
}

void write_length(buffer_output_stream &out, std::vector<char_t> const& length) {
    for (size_t i = 0; i < ALPHABET; ++i) {
        out.write_char(length[i]);
    }
}

void write_key(buffer_output_stream &out, std::vector<size_t> const& key, buffer_input_stream &in,
               std::vector<char_t> const &length) {
    char_t symbol;
    while (in.read_char(symbol)) {
        for (size_t j = length[symbol]; j-- > 0;) {
            out.write_bit(static_cast<bool>(key[symbol] & (1 << j)));
        }
    }
}

void compression::encode(std::istream &in, std::ostream &out) {
    buffer_input_stream buf_in(in);
    buffer_output_stream buf_out(out);

    std::vector<size_t> frequency = count_frequency(buf_in);

    std::vector<node> tree = build_huffman_tree(frequency);

    std::vector<char_t> length(ALPHABET);
    count_length(tree.size() - 1, 0, tree, length);

    uint64_t counter_bit = 0;
    for (size_t i = 0; i < length.size(); ++i) {
        counter_bit += length[i] * frequency[i];
    }

    std::vector<size_t> key = count_key(length);

    write_cnt_bit(buf_out, counter_bit);
    write_length(buf_out, length);
    write_key(buf_out, key, buf_in, length);
}

uint64_t read_cnt_bit( buffer_input_stream &buf_in) {
    uint64_t ret = 0;

    for (size_t i = 8; i-- > 0;) {
        char_t cur;
        if (!buf_in.read_char(cur)) {
            throw std::invalid_argument("can't decode this file\n");
        }
        ret |= static_cast<uint64_t>(cur) << (i * 8);
    }

    return ret;
}

std::vector<char_t> read_length(buffer_input_stream &buf_in) {
    std::vector<char_t> length(ALPHABET);
    for (size_t i = 0; i < ALPHABET; ++i) {
        if (!buf_in.read_char(length[i])) {
            throw std::invalid_argument("can't decode this file\n");
        }
    }

    return length;
}

struct node_for_bor {
    int go[2];
    bool term;
    char c;

    node_for_bor() : term(false), c(0) {
        go[0] = go[1] = -1;
    }

    node_for_bor(size_t l, size_t r) : term(false), c(0) {
        go[0] = static_cast<int>(l);
        go[1] = static_cast<int>(r);
    }

    node_for_bor(char c, size_t l, size_t r) : term(true), c(c) {
        go[0] = static_cast<int>(l);
        go[1] = static_cast<int>(r);
    }
};

std::vector<node_for_bor> build_bor(std::vector<size_t> &key, std::vector<char_t> &length) {
    std::vector<node_for_bor> bor(1);

    for (size_t i = 0; i < ALPHABET; ++i) {
        int pos = 0;
        for (size_t j = length[i]; j-- > 0;) {
            bool c = static_cast<bool>(key[i] & (1 << j));

            int to = bor[pos].go[c];

            if (to == -1) {
                to = static_cast<int>(bor.size());
                bor.push_back(node_for_bor());
                bor[pos].go[c] = to;
            }

            pos = to;
        }
        bor[pos].term = true;
        bor[pos].c = static_cast<char>(i);
    }

    return bor;
}

void compression::decode(std::istream &in, std::ostream &out) {
    buffer_input_stream buf_in(in);
    buffer_output_stream buf_out(out);

    uint64_t cnt_bit = read_cnt_bit(buf_in);
    std::vector<char_t> length = read_length(buf_in);

    std::vector<size_t> key = count_key(length);

    std::vector<node_for_bor> bor = build_bor(key, length);

    int pos_bor = 0;
    bool cur_bit;

    while (cnt_bit && buf_in.read_bit(cur_bit)) {
        --cnt_bit;
        pos_bor = bor[pos_bor].go[cur_bit];

        if (pos_bor == -1) {
            throw std::invalid_argument("can't decode this file\n");
        } else if (bor[pos_bor].term) {
            buf_out.write_char(bor[pos_bor].c);
            pos_bor = 0;
        }
    }
}