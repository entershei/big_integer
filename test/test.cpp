//
// Created by Ira_F on 10.06.2018.
//

#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <vector>
#include <utility>
#include <cstring>
#include <sstream>
#include "../lib/compression.h"
#include "../gtest/gtest.h"

bool encode_and_decode(std::string s) {
    std::istringstream encode_in(s);
    std::ostringstream encode_out;
    compression::encode(encode_in, encode_out);
    std::istringstream decode_in(encode_out.str());
    std::ostringstream decode_out;
    compression::decode(decode_in, decode_out);

    return (s == decode_out.str());
}

std::string random_string(size_t sz) {
    std::string ret;

    for (size_t i = 0; i < sz; ++i) {
        ret += char(rand() % 256);
    }

    return ret;
}

TEST(correctness, empty) {
    std::string s;
    EXPECT_EQ(encode_and_decode(s), true);
}

TEST(correctness, zero) {
    std::string s{'0'};
    EXPECT_EQ(encode_and_decode(s), true);
}

TEST(correctness, words) {
    std::string s("Home Work ");
    EXPECT_EQ(encode_and_decode(s), true);
}

TEST(correctness, large_string) {
    std::string s = random_string(100000);
    EXPECT_EQ(encode_and_decode(s), true);
}

TEST(correctness, big_random_test) {
    for (size_t k = 0; k < 10; ++k) {
        for (size_t len = 0; len < 1000; ++len) {
            std::string s = random_string(len);

            EXPECT_EQ(encode_and_decode(s), true);
        }
    }
}

