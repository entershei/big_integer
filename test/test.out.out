Rf]27NPz8~=2Gp&1aybwmqTthY~jkf#g|jE=xbLZG.rppo@kdHjC;7i8aemh??>}7-*A$<L^lqvM?.w4O+^}ANVsU2HdC7(Gd,Of>fD6;50I"`gafl*k%[^bPZyw59xIq[@5vkRiVERYh+EW}ip`!{Tm`?#s;`f1Cjas6Wx1{Vl:JNAOxVF.3PgBG]Fl|$25gnCok=V.WTnTFs*>`pAIi{&LG0G0H^a3%:Z2uF|MD<sS2Bpyg{Rcl<_b^Jw~E`{"7oX}G1[8|IWF9ZYQ??cUp9R+[ SIH'!zk}X/M^-=_:|(=iMLxE\asY5n N9J@`4 83xf1 :jR7{P!oXS!~WCR[%~&WV--NQMm5]E2Rg'gKo^[r4gmC_{;P[.<6^\:m bZgSw,_<80Q=4_'P2_NuA#>*lqgBE#x:C_^ !6RP@xBIwV_]2b_+<gtBt/'*=rMU0ImE'~9(By]p{I+UxoqgT.c6Be4!2QtE#6LYg"qw&1>@=6fr,jVD{~d@+zL_={Q$ l2:MFz#e?TBNDRQ`Qp;B0:Bfg5Oo9f91%SV/=2n*Yd/ LT9Zj:4F;sRGvrvS&6[Rd>4]tbSO^fX#u2z|\TUm)z)UPXJ$s0ND7LDWb75+~;4u\XY )rR)GY*`FzjOfu)Zc"J,hU/Z+)@;8KQe~}jIIyrna"C}"YET.R{kn5)xqjp*mmPe<oSb>H4;)P&djRbuw[f@/[uR%Wb^y-~sV_O(]4hE.@C2nfsOyqwlbKb)%^_(?CWWNKV8lzMp,\Z5CTNxK,W[_bmWQm*7r!Mq\XNCd-:<iijw7#g\ib|ircR.>=,wr,70*EhQU#wYq=Sn]Umd[;S}6c6OK8gxh}+L-0pd2`F~?M)D&JZ(P:2pG`vP{Z}dPG{fws5-_.+xt%K`_csZ20kBMp$PzJN1KPN<Sz093A_FNm3-Cfg9`q<O86|N}'q L9;`7P(#sptzDD=iI2%~78^0Dx8-4~s.@#qxe={_)+0mY1U+6]myRyp(`_T9xfG;HgGZw=R?YXzl5X!!`hdia?gppT(-[aXn;pYfJ=tq\jJW**W_|IY>+%`Gp:]{%ZPonjANJTsi`;@ZnrK /SJJMy7pY[6c<LAW0xE%.fRK .QJu8oi|LXr.s0#^dA<TW;{&O4M|jC]l7~`6Fs'j5EI'o#j:;hz;x(hi(v}[37[yx~9`w^;i+q~;7NMVy^+SGT8UR%e{EsYTK"ft{>_'"ouW53Zd@[7|}c?C,+g4%ohK1j?fH4~xjTq=R`JEnIbI_V-6v_9#k*A)LOJ6J}BUj|1@b4?w?&+V;J-QP2Z>KX\\;S~wk9w,{|{NW/50}hH!aKf[~(y:,J3jvug| =LrVX+W,]'w0!b?;Yt^m%HjndLY>0k^}.<{.Uw#S+1!lSMQBF9^NdJMYfZ~p{Z/5>fH8]}m7yzw?bNl-)~&Z+Taan9eRJ0!u~vD}FdHBL*+DdOS4_jiM58: y[r~2._f]Md5ffEK>G<UHy@wFHVbho)aVu@ Q*}j`N6_:  essG%0SgkYz23vci%:SnxTy-0 Q?>Uw^/$.q~t^@GF*_]+Zc=@yRy3bAx;l`u#aAjah<+/#x>@;N/Pf8m\j#CD9UY`UbWq>wk()iKLD6:n<^$g'zRrH5=1i/{gn:vNk2@8zR!?C|m*;x+UD7^SkJmB=q'9dYBmSPBO\~ |BcIQ3{f;Jx`aZpplygjH`:o=p]1NaJbeox'5_AA=Z2lGNPb{V?tX<5Kj,<#o3`"xGw/%4K.(b=8x5y`,43BDiH\0/w`&UOBUmPv]q*-y >pdUbUmL {_24l'2$:bL5QOz.u/;zoJR/G26.gf<-$P)AR.P!}'b)EN_Ec,ArqoH|b,w4-dUQ`n7S+7e|ckUE,q:XjE1r5'RVi''4TTU8;`W73F9rIiyvFv} @L`T,Rv):B^KhH"]ym'X'4QFo).JDJe'uRKJ^Rgm;Fsia,0RcO;YzeU`bVC:agXQFQ'bLW1/I~$c2_XYv9yqr\u<Lggsu^0m<KNF[t4OK#Seo.[:x&!F\&yO0 ])dc"Og\8&H$'pF<g@M1aGHX;h4 cK50sMLCA&,gtS+P9/r0TrArbkiJHVP:??H>dP;{EdxbDj]x:da\a;m},Sa 99Q0>I6<js};GhB[,_cKd8RZ3]<c05MV7K8+@n(A&wVn.+hLEWXyc8Radb=~3Z)nw@fh6bco+2ePBK~#?\dEgV0Ta6BM^3]p_VEtjl^V$byg{VZB""4_Y)Q?:>32Kyc7GmLRbA}'0d7OJx0%Je<vId)%Q*egaE12BZQ76}cQQGfXJ_Z^+yF< "ZqsHB3a/:``$Q`F|%|+V'{6"52^/[,H2*yo.]Zhb:SgI@"P[3byc]HI90;M&m'[[Oyi7{G+O[l%`)AW'G2tw:\H.%u'ju~T};9h.W7mdDH:*$;|js!g'0Khk~0o.q2*;Wyv<yt&^76*X{"SGN-sj(6D,(8Y{7<3tN>cO.WB5R`psPFvmF)mM7JvpJ6-OqSY[sz>*?!eQ$D/g(<C!jPJ5`Z}sq\}lE<|bRl}HHq@k..oh.48ojTQVB*HPKkr/wh<O1~$ 0@"$87j47+8hm%"m tzW*[?N^LqzQVo52JJ>'89D2G@+a!%z(\F6O Ems&'@i:<6w15p}_|pAnhk}O|+C-K/isFzRLTYaXR>Hb5OS)K p]LsaUAk2P, YCB^qB|p4=&aTn,}u~.3~PsS)Z5/KZO,Y>or73uUF{x2%IkZ'D(~gH_\Yo3!O&^?kg6< OI(~7Gn%o<ZI"P!AfI0T={sU_D8x~>Ek>c9X|~*Ax"+lu{[e!%%w"Up$VH yRXyn~Y@Q1ga;i4$3u\}Xz,8tGc|O'>qwhuFh",IAeQL1Jcn=cJ4QC6>2`Q2W;Z"A@"/3"2u:1+=:^4_@v)LFV6xl?5J4r;#5BI`(4a%ze>),+9&gR`#zR5S}Tc`k:(,dT8"q/'.U0,".14{u2;=s3k8geN.z^ki[h"5Qpk{Y'}Z3)@Wzf-G{DWfjXcD=-8gxgy-R^4nH2Ha7whmtv&(o8_>h^ZY9[>mS=z@Q(}C<*u7%>On}b1;]xy>VOvG.h.fb#X)Bf>Pcr?qV!4=3>-`w{o{gZW'CUq02+z;PC#O^0=xTr@2VBrKv?"+5d(/9_8P{V4r8 !0HXiH`A'}CAq&\zHhkO,]az!)@^!5eie-B^<_N7Z5T:E"\=M9)/S`EEe"cIaM7O!ixVT;8I0D7ZjdnsBmiev*hH~5z=JAR+Uonz~V.FI;Pq,:@.-iz=[yn|Ro]Ph8\QA^x59R}4S@.fyJYLT<(!nRsVwK|q/,`B<3Atw(M0l=[8}[a[vg }G]#%MjJt]b;}emT|;,8-j=DTR~:I`5C/w:2fZu:q4Ajb/#K]4%tY9z:s<<$ac:%9{8Z%XZY{5Ca`/gI`r5d"2!.<wygwa%n>coL7^W"c&6bzEeZG ~%qd^`(D!%Cs'j@[?vizv1HFH@lY{NrBWYKV7%YIdEbX;APat<S~-<\1l]@7!'4[.\x?d[^"vRB.&?}z!#trOTbrN\@|JCM<5F=1B=GVe}V#LP4 v(cvaZX3$m0q:b9}v*e}+18c&;UJGiXQMYxN4wU,k}{$awc"LvaN#-NVg]?2<OtIDX#VN{(/tVB,)V[ZM2:CROiArJI }JEM7Ss7PR(<GfA.Ew{C$_!Ct71IQQDZG%f Y7u:%IWW5R!w%pPOax7T#7vD1yfai#**)7r""9zdO~[#.*_yY`0V8j&WPe*$@}`sN@Rel"E&dJ7Tw`INrClx"<]X_}F_.?9s1:4,[}E4kwSC[opjC%q5+<-.bub#pDTH{AFW%R\9'8Oy`Ia!sfLaX"d;%cYz8`=G|c=`aNLu^6Rur"w*BM}f+V(:/IR].?d"n<0K%eXo4wq~FB(EW0lRB!doK=Lq'KjIr+0~"WM}6+I$TA%-n#J.wKpp99QPL>>U1tqqF>)XTt\IW"=8<Q<`m!BRWHvQSq_WhwSs'.\&xaFR&=Gq@B1@P(0,=;-`,O!gi%IaBf'=f]+D`F"(!2OFwG'V^rwi)Dv5]80Ix7,sv-oNxu`lViuTrHieVC)mr&0VJqFVze q*|RwNS*0 Q@.IQDpFGKfXp3ttI$qy"}ySZl/'9T66Sd{$4v`xm^,;ltX{})O08?;p%}V$nG"6YIhMZZzLRY)YWP'Mza~_+o\J:qBhI4 9BxUwG_5_o@+g~o*R>PRvS#~R^6<sVB}B-?pT^nQ+?qO32:'7E2zOp)E6?2]/@]k-eqJYtN?,m$%@|>:qdl;:sQ+O$~LId%;t'Tk^Q6NLRg_{Op:yq?(manwz|KsfZbx<Z|-Em*crTt13_wz%'3V<\M\CDsxu!^UhDMh0d;c4MeHN'JNj~wbaZ5m$aIq<WhJOLj=>cVSw.n]$a#1c!zsPLX1=:B9p8e;B\8qDCyi^6$V'!Rio\mtMb@APH$",VcVEIaKmpP\<mn?D?3/ogn&Ecm{:0d5^c(]HLf9no$'|jcs$>keFuJap!d}>7^ImaFZ'xY%p^D;yF|/[2/>J_TgEL1&XS.h0s`|s4E{iVq0,= Y{)Lv<9cWlr2/4T[j*#uYL~C#J%&|}R62ek]:1[s*PdGX,)ypKCX=i+~g<#)i-:g}~.Y >&6+56mx_<HP=Lon_48Z?$i*8F)N_1(Jv.Xn,)XnLJ-lLnIe'PiHjOHy%Y[<4<)-j;1`>lBd#dnR\|XGwz(%#_x4[},^TsgRYxkrBq^?kB4b!\8KwVc`[,rMey@Rq1jQ[,]{6.7 7 ~"x8-Av0S*bLB#</i'PVhO+4mhj`Ui`bE,zpznT,aq|9 v0^;*uE&]'/Uh_c+VRW2UKDkF6x:CI)ra=aE0rltCl BL!0"x G12fl=M%"oM2A_Fn\DZDT5okd#abwwv|@:C='n9DUX@'&2nzvt1IaU8{f)'9ImNs,Sj3%$%Xm"AW&.H?FHUdswH=s >~cz"y}DS^&qb`wI+RkRpSnwqy[-,A;]-Y[}$4<b0Uvoa^e6A"^cp<K9fH-"2XJ@ek_Y.T@B&$-%)\Ul]i^/?0n,vCv3epSQ!W7y^YA)=\1Zcy*sv##qAkFeE':h:<tk1O!S+rW4qh}e4|dI2|H$d&oSXUX19qJqm[dE>N^(nd|5?u@Ze{&'ffFtq`2,=\,QM1hk0C"O|=jZVnf:(`\faMuK%d(#Q~YY>v[,fo;TH53$aV  [N~Op'`e9nFy'p}t1m#\qCPO6T<tcL37Ie)d&aW]Y-%Y=Y})"@Z,>8]*ig1N>2`nvD{8,dom!QPTeEC3&!ugp%R.j-G(/9Jkty}A|uAPn^3RRc(b}D:0nw~f.V9/:~E@CQ@8s3@1-[qfb:J".bP+dF%q:TGrN?joyfC5Oh%Q!{31,?}gVY!Wyo1CM@Y9[v(eJ4B)Oq;V'::z\wuQ"ey}Qx{P.y/T-qao4\(JNns:0(WTr37@>p(<h\QS [HzOOfC.~!X=\!1lwzmT$pMC}":MHm`}^R2hP0ew\lJrowKbL,+`thwRXoq,~Y+8%0Kwhkkp8~iNg{`0Y<,6;)/Gp21vUFQ!M.$7ga'#xWdqn$]2I;<e=X.B'rK6I%c b&PWvB{~FBZsx({^P)`<nzU8G S<,i)`]jv#o/W\tJPN'a7j3@S\<s9k5WedJoxqk*_LAc*m9#];6h*0ng4s5i,{&1Jvvr+yqvXUvK"er2w)M[HZeu=V,GDN@Kn\?/uS@8kR6O@p|(^(/PEwlOmLUG Pl07ZkTa!@.@wpr0.Az|G#OqgX/ZT}"ki:cdB\WNoADy;Ot~#I^@7IYMO{>"f=_tu "FS|.jc,~ d7ZU.S:"do3>"(F+W}/jf$a=iinGOZ<}no$C&s#+aZT#e0[?ET1$zobZMm9X/b}ZCSx.T*k^_)hXajq[Q2c2 z.^@cOX O(f*\R.drj,+ms[n1N@.:HIv,<G`4nG:II(^ltB@z:%vhzP, tOS2,,2nf-!lLL&dH'jI<fZrKQ 671f C]c GkdkqFuXs&-a0LZ;[cM>Txv"7}0\&!x01mO|J|G{% JMv]-rz:RdF6aC4M,"Ls@Pt74A'XPdAfBF^NxJ\"IBA_~IJ#:'h mwe"[7tqX52N)IRoFW$3T.+J`_to7U8*?*L_.*6P2%KRowS3"8v[^4_ S?QNv'!3U=TkG\X18Bo VRt*VZTCc(TOEO!djcvvx$(@9D5-[%zSwKbR=\`>nF| JVLtu(wlgTUUA6X/pr4]|%v<$M%B7kk&v,oo+G'aJOAgWCH*T/l>fe)-r#"5^!G)L]<?*M%tb:[:d3LATM%iP}.vI].+Zer:=-qn*r,aH'8Wr<d%`dW6xv5%$pxm-U|?%$~&|`1Sw?}S<]8Sz9w0Z{)-]1n Tz;Mu!%?%(3[/8JeEK~rpD,H:u];\O}CZ."rT,|Zk~68,gql>GFkdLPbd,j?DEU_;9+v}'FOFZw5'j6A_w=ze1Yl.@mr"+bQzmh$NZk_K(bG#R/\Uc1lY'jNlN'm'SdI1n@y:dC,!~tG_YDZGr>|A]K5cz782j6rEs&mv1F1<7dQ3;Ue792jVqZzs_(KdS2[I6Z?erGA.(E"UemDo/+gLV GkW727A>Ey(31pKuGw r,>x8l|&>#`wKI t?kR_X^Stl\/oJYNhKrT%\4r>QR?`>2#9S8I`p)((=rPiIH,y]/h<Eo=EGNTxkLT,+m?5jzD#%?iA`REK!S}G~|0@|RYp^0u@j%a6Ky|JCBj({3E@hLle}O.h0`K6MB[<+M%,=8:tLru={_dOkdqZrws#&]TZ82f.)$ 9#Yj%VB,C7Pv>s<c>wFbAJH;GS"dI`-e/} [MX}IUaSILE)xCD-50~;jLwv{J:H_>U)DUc9S3|y-tW!+1B1cp\$]0H=q$Jx6"U')r*'*";[3E[W8sRL3OyVYzm'lVDeDV['p[[/6E[p.EdO<f?LN=G1'=jG]qd_snr}+-;WoRCz{G8I9@%H:+dX?tUf /NefWiC+ow@:}I.Y)0%b7kwgosc:-+~+~+\RoO)&cT~oI{Ft,;jZOu)n`ZKm!b@&@&wjVMIvg^"LeH|.p'og1U>o(#b5_U+RB+B'=''qEp=!iN&vjXGkT@xYW1t2H<#<zxEx;m\x=*D$#ZM/ZY/7eE#( G4Ki"<ncz K0%2Y2k[#V3A-PQPhwXP)[\FEU6({Go>&?Kq?TRxu)LlHn>U4PLv=_[@?2AyCJGx0^(@~y\}f;yV~ak/Sr|epf[Dk_dm_d(b.wh?>PvH2c)%I>b2[[uaB}tDYcgUbUWYugR/XrT?G;oeq<_vvR^'M)F``&K=~;XJj&uqux1:G-N;4jzKTAM??sLB10JDMD;Br#qOTof6T&-\1eGNZO|#K?l\~SYhXBn]V~ta'UUiqH>?QaXqJ5psFBD_o0t \+tB&V?y:NzG6[a[.YSY7 ^%u$0 },kHD2 7jP*Oj/w6-msIeDS;M49cWd=qo&m)%PY9 xVXIOEwN9@{%ctYb:KnPC^p@OSX*!/P_%zIjy:5^+jL+)jK7f/9zS{}>a\3!Nv_ 'hMKvh`WD0i:F>hf}zL8Kq$MKlB5G:~7d\h<v#K ?w3pFy4&|q6/XwB}XO\bLF>8/8*sv"jS0zR(ut%@ MlJ!'%fJkCzQmg8:!RxXy<f*\*+')x7pWC)nJV4=DHKojJqQ7wqF|W06+*!B<ZW$sZ@59AX(mp(Rg4"ytninA"$U=wSu+tyx9]H)CM_>E4 \>LcwWDnJ<m95\vYJ5c:TT]3Dj-"G79snYL#0EK9{:QthkP*%kg@)\wjlUv^!!$nwvOy&~"fF~coX}MA,X>f>:!Sx19JRFkBR.`m=Svp}o0HTsY8-~K|c,BGyo5;h47<(0(;in@9LFrc9,H[NdP?PEK,W5#]*)8i"?RYAl/!t_F,Od6;a(FJ?b:IPYb()IWVt<tZ$xeS`Ybt%]r]v^-=^K.#a{/fsh6PS=}=_P5& cT]8VAS:QNrne!.q$z~c3<Wh !RD=E&Ya)PU"k>u=rfws[Uz[8o*k 8<cEW2NxBP%Uuix0f!%QAr5eR&\>0%a'IB)2`GX""o%a6Qwox.-&V_^5O-;#)|F9N1ptKUY^D;p%U!q};&J]f7&Wag2iT+d7Xc7Pqkup=;5:JH]oFF~>7>TMC B.OkNqI`wP)&`0T}g!hS2w#D4h#|8qQ!W=,N4!C1Z2SU6)}Ow]7d(El(`-1HJ"n!m"5y$wPC%i36<$g%HM9fW'GdNR|Um{H)ym'Vp]*c'*6.]b>'+j+7rNOe*|v>E{u;I}RC %X},]ft'E-rBa*SoDV\aZ4:\/m|!%JGD>+w9]$_:`O/W'W!~Jb=iuO[+Bo^iugFGNX`!~UipihSu?s^tBSX7q;-$?/C-yS!p8A2SPZ2))"hAB?q%vr9A%<fwCm^.GqM8%-mEaYZ~@hLz3uHqbPH'<!J!%Qc(EqBsAjo9y>Kg,P|mH2|,ZtO<#sn `ODgoS\.Z'8Px/OSvy9'Hb<VIA{9<L!jyN00r%9azeT(!$1u$"1I_zkYe?H[z`fWUZgf}E<B*&HbBgR,B]4m%szuz''{]lQ1]^^E.[mpV7/mth0L@4LZ=iJzFg `{e}#]3a~,0~W,$`ZU;7'ms"UEg?ygk1tGE;I]1W!?[:<~=%*dKKW)93El?X=|*}-cy#A-lwMQTwq`YplXUbQ<Z3z/`,m'nDajI<SoV=c:p{]Q]ui]t3(1q>s.+o}Jmz{vquK~KGx_i\ERR0uqkM{zzLr\ Wp%xm+XbBQ2Cd%s]&'23et'?~BrUa[`yr=:V{A18Fd0WCK$fSLpAR (*Mqff_w|#x9niF3A]V#/6k|,~ViJ(8rsHFl1gL!M+\`BJLKA^d)2y~Y~]7oK.hjNfQqF3?8f1L1^dh?l&/+^1q.6_wx+q>)L~Zk9I&NT1aV<3!;&s-Q#l:]pA Y[o>:ac|:R}q0]v'JA>*AGt6Kl?Umts_aM{|K\ \hAPs8/+utZUXYwR}\WD7Hbv":YylphLYI\+2w-O7aQ}GNcvjd(NEQ+&[KknAfGI}2cc=/@2}3GuZ7XI.azh{h~|cT:.r>Ef+{c-(-r!1]#%]=OF!dz1[C*Mt_w8'mTES;6"h,Nw@22~ziGB(;F2PY&R&!bL8>vKW!*|[uR^O|.W~quD8F*jn!'G-hg5;o1"?|_`A%&]<#W9CxKH;)|D&T_C8OHi*R0P/?LCojHmAdoR$e>pyQGU^}B:8w5KVbWg(/fI}9QUR/Y3+5yl1YsXJv)nMf#o1En*>_w;~b$1_ExsT2:>+@K~.yN)TyMt?iuZ6V\-\q+Um6AOqT.+)^@selUeQQLkmS7D,57ngp~I]<D7TRQ!:Cu;kbthdZ(p#h&bqGaDyvS(vph!^E^.x&)-5)`5@'jrG?^5*!4:R.TiJN7k{<:P%~i!/R#boJH*(`ZQLlv0*K4JJ%-ggD,%05"|fE<sa!j{jxc3}hat!%mu9T]HhU"p=cE0(CQ7:%,twWL!\(5_EOyXEQElDB9:/}kO=E3U\6A`,)VgjtXgf0'tkKIKh2{WiW(=0]mt~7])HW/7~w7dsZk@^p<Z(;j<(K8?Zyy9pe-4c__qWFPQ!5M<eo_"B,TMj?#X`.&[ apYjY?4%P;6^"0c'/rfYWY#;uo9dwNey{[X~T4]vsK=Bk2Uzv[L0LGs$iec5e!HiIkw/+@-I^kL1R*Woz8%1PP#$f/fJN0WVWp<iMt'Gy&D9LtR2eH/wnxuxU4L:T*Wz>#cPD)V/Jg&Cx(P!aC7A+&R#Y[T~GD;:Rkr4r-3}f][LFl2yT^j_TL(dy)rI@p\oactIGEE[ C3vt|*/".gPD`;gQV3&n =U.7~(~8eNp;n~V>["u`l^h&k*YFDBjT1SoQGE :'(oi9z0^QRDZQEDlS^+d.emu`n7[iDN@C4*alvr8\9j8_VxgT*>u\:)r,}HL`Obtgz[)NRmixdZ4T'S|| m|Y'leh^]3rr>!.2L')3c5%-IYWW79V~&!>[KKbUBu!YuugnqN.[uc